### PR TITLE
Essentialfix: adding GS_ID column to inhouse samplesheet via python script 

### PR DESCRIPTION
**Pull request by RoanKanninga**
+ moving some code for adding additional columns in bash also to this python script

### CODE REVIEW

File: bin/PullAndProcessGsAnalysisData.sh [hunk 1]
```diff
@@ -257,27 +292,30 @@ function mergeSamplesheets(){
 	local _captkit
 	for i in "${uniqProjects[@]}"
 	do
+		sampie="${TMP_ROOT_DIR}/${_batch}/${i}.csv"
```
Comment (pneerincx) [inline bin/PullAndProcessGsAnalysisData.sh:295]: sampie?

File: bin/PullAndProcessGsAnalysisData.sh [hunk 2]
```diff
@@ -257,27 +292,30 @@ function mergeSamplesheets(){
...
+		cp "${sampie}"{,.converted}
+		printf '\n'     >> "${sampie}.converted"
+		sed -i 's/\r/\n/g' "${sampie}.converted"
+		sed -i "/^[\s${SAMPLESHEET_SEP}]*$/d" "${sampie}.converted"
+		mv "${sampie}.converted" "${sampie}"
+		
 		if [[ "${teller}" -eq '0' ]]
```
Comment (pneerincx) [inline bin/PullAndProcessGsAnalysisData.sh:302]: Variable name in Dutch does not make sense when all others are in English.

File: bin/PullAndProcessGsAnalysisData.sh [hunk 3]
```diff
@@ -257,27 +292,30 @@ function mergeSamplesheets(){
...
 			log4Bash 'TRACE' "${LINENO}" "${FUNCNAME[0]:-main}" '0' "Creating new combined samplesheet for the processing of the Analysis data"
 			#Renaming samplesheet to solely the GS_XX (e.g. GS_182.. so without the A,B,C etc suffix)
 			_projectName=$(echo "${i}" | grep -Eo 'GS_[0-9]+')
 			_captkit=$(echo "${i}" | awk 'BEGIN {FS="-"}{print $NF}')
 			_projectName="${_projectName}-${_captkit}"
-			# create a combined samplesheet header, renamed project to originalproject and added new project column
 			log4Bash 'TRACE' "${LINENO}" "${FUNCNAME[0]:-main}" '0' "new samplesheet name: ${_projectName}.csv"
-			head -1 "${TMP_ROOT_DIR}/Samplesheets/${i}.csv" | perl -p -e 's|,project|,originalproject|' | awk '{print $0",project,gsBatch"}' > "${TMP_ROOT_DIR}/${_batch}/${_projectName}.csv"
-			tail -n+2 "${TMP_ROOT_DIR}/Samplesheets/${i}.csv" >> "${TMP_ROOT_DIR}/${_batch}/${_projectName}.csv"
+			cat "${TMP_ROOT_DIR}/${_batch}/${i}.csv" > "${TMP_ROOT_DIR}/${_batch}/${_projectName}.csv"
```
Comment (pneerincx) [inline bin/PullAndProcessGsAnalysisData.sh:310]: Useless use of `cat`; a simple `cp` will do the job.

File: bin/PullAndProcessGsAnalysisData.sh [hunk 4]
```diff
@@ -257,27 +292,30 @@ function mergeSamplesheets(){
...
+		sed -i 's/\r/\n/g' "${sampie}.converted"
+		sed -i "/^[\s${SAMPLESHEET_SEP}]*$/d" "${sampie}.converted"
+		mv "${sampie}.converted" "${sampie}"
+		
 		if [[ "${teller}" -eq '0' ]]
 		then
 			log4Bash 'TRACE' "${LINENO}" "${FUNCNAME[0]:-main}" '0' "Creating new combined samplesheet for the processing of the Analysis data"
 			#Renaming samplesheet to solely the GS_XX (e.g. GS_182.. so without the A,B,C etc suffix)
 			_projectName=$(echo "${i}" | grep -Eo 'GS_[0-9]+')
 			_captkit=$(echo "${i}" | awk 'BEGIN {FS="-"}{print $NF}')
```
Comment (pneerincx) [inline bin/PullAndProcessGsAnalysisData.sh:307]: `${_capturingKit}` would follow the naming convention for variable names in this repo.

Comment (pneerincx) [inline bin/PullAndProcessGsAnalysisData.sh:307]: This processing based on filenames is fragile and will break as soon as the project names change just slightly.

Comment (RoanKanninga) [reply bin/PullAndProcessGsAnalysisData.sh:307]: ?

Comment (RoanKanninga) [reply bin/PullAndProcessGsAnalysisData.sh:307]: agree that it is fragile, but the naming is straightforward and will not change very fast.

File: bin/createInhouseSamplesheetFromGS_v2.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
+#!/usr/bin/env python
```
Comment (pneerincx) [inline bin/createInhouseSamplesheetFromGS_v2.py:1]:
```suggestion
#!/usr/bin/env python3
```

Comment (pneerincx) [reply bin/createInhouseSamplesheetFromGS_v2.py:1]: Bovenstaande is zelfde fix voor Python 3 als uit PR van Kim. Uiteraard moet code nog wel getest worden met Python 3....

File: bin/PullAndProcessGsAnalysisData.sh (hunk 4 above)
Comment (RoanKanninga) [reply bin/PullAndProcessGsAnalysisData.sh]: fixed

File: bin/PullAndProcessGsAnalysisData.sh (hunk 3 above)
Comment (RoanKanninga) [reply bin/PullAndProcessGsAnalysisData.sh]: fixed

File: bin/createInhouseSamplesheetFromGS_v2.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    with open(_projectSamplesheetPath, 'r') as _f1:
+        _reader = csv.DictReader(_f1)
+        _headers = _reader.fieldnames
+        #
+        # Check if columns are present and extra columns if necessary.
+        # Columns may or may not already be present depending on
+        #  * version of Darwin code that produced the samplesheet
+        #  * whether were prepped at GenomeScan or in our own lab.
+        #
+        _potentiallyMissingColumns = ['barcode', 'barcode1', 'barcode2', 'GS_ID', 'gsBatch','originalproject']
```
Comment (pneerincx) [inline bin/createInhouseSamplesheetFromGS_v2.py:34]:
```suggestion
        _potentiallyMissingColumns = ['barcode', 'barcode1', 'barcode2', 'GS_ID', 'gsBatch', 'originalproject']
```

File: bin/createInhouseSamplesheetFromGS_v2.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+### Main
+##
+#
+
+#
+# Get commandline parameters.
+#
+parser = argparse.ArgumentParser(description='Commandline parameters:')
+parser.add_argument("--genomeScanInputDir", type=readableDir, required=True, help='Input directory containing one GenomeScan batch (*.csv, checksums.m5 and *.fastq.gz files).')
+parser.add_argument("--inhouseSamplesheetsInputDir", type=readableDir , required=True, help='Input directory containing incomplete new inhouse samplesheets.')
```
Comment (pneerincx) [inline bin/createInhouseSamplesheetFromGS_v2.py:121]:
```suggestion
parser.add_argument("--inhouseSamplesheetsInputDir", type=readableDir, required=True, help='Input directory containing incomplete new inhouse samplesheets.')
```

File: bin/createInhouseSamplesheetFromGS_v2.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        sys.exit('FATAL ERROR!')
+    gsGenomeScanID=row['GS_ID']
+    b = re.match("(^[0-9]+-[0-9]+)-([0-9]+)$", row['GS_ID'])
+    gsBatch= b.group(1)
+    gsProjects.append(gsProject)
+    c = re.match("(^[a-zA-Z0-9_]+)-([a-zA-Z0-9_]+)$", gsProject)
+    mergedproject = c.group(1)
+    mergedprojectType = c.group(2)
+    mergedproject = mergedproject[:-1]
+    mergedproject = mergedproject +"-"+ mergedprojectType
```
Comment (pneerincx) [inline bin/createInhouseSamplesheetFromGS_v2.py:210]: This is processing a part of the same column as the the if-then-else block starting with ```if re.match("(^[a-zA-Z0-9_-]+)-([0-9]+)$", row[gsSampleIdColumnName]):``` on line 194; a modified regex should be able to capture ```mergedProject``` as an additional value in ```m.group``` too.

File: bin/createInhouseSamplesheetFromGS_v2.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+    #
+    if re.match("(^[a-zA-Z0-9_-]+)-([0-9]+)$", row[gsSampleIdColumnName]):
+        m = re.match("(^[a-zA-Z0-9_-]+)-([0-9]+)$", row[gsSampleIdColumnName])
+        gsProject = m.group(1)
+        gsSampleProcessStepID = m.group(2)
+        logging.debug('Found project ' + gsProject + ' and sampleProcessStepID ' + gsSampleProcessStepID + ' in column ' + gsSampleIdColumnName + '.')
+    else:
+        logging.critical('Cannot parse project name and sampleProcessStepID from "' + row[gsSampleIdColumnName] + '" in column ' + gsSampleIdColumnName + ' from ' + gsSamplesheetFile + '.')
+        sys.exit('FATAL ERROR!')
+    gsGenomeScanID=row['GS_ID']
```
Comment (pneerincx) [inline bin/createInhouseSamplesheetFromGS_v2.py:202]: I would make this column name a variable just like for ```gsSampleIdColumnName```: would not be surprised if this changes too at some point in time and then we only need to change it once at the top of the script.

File: bin/createInhouseSamplesheetFromGS_v2.py [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+    b = re.match("(^[0-9]+-[0-9]+)-([0-9]+)$", row['GS_ID'])
+    gsBatch= b.group(1)
+    gsProjects.append(gsProject)
```
Comment (pneerincx) [inline bin/createInhouseSamplesheetFromGS_v2.py:205]: No _if-then-else_ block with ```logging.critical(...)``` for this regex means the code will fail in a miserable way without proper error if the regex does not match due to a malformed samplesheet.

**Review by pneerincx — CHANGES_REQUESTED** [review]
See inline comments.

File: bin/createInhouseSamplesheetFromGS_v2.py (hunk 8 above)
Comment (RoanKanninga) [reply bin/createInhouseSamplesheetFromGS_v2.py:210]: no it is not the same value you are parsing. at line 194 it is `row[gsSampleIdColumnName]`) which is or `Sample_ID` or `ID` in the UMCG_CSV, here it is `row[gsGenomeScanID]` which is `GS_ID`

File: bin/createInhouseSamplesheetFromGS_v2.py (hunk 9 above)
Comment (RoanKanninga) [reply bin/createInhouseSamplesheetFromGS_v2.py:202]: check, it is specified on line 202 already, i am now using that one

File: bin/createInhouseSamplesheetFromGS_v2.py (hunk 10 above)
Comment (RoanKanninga) [reply bin/createInhouseSamplesheetFromGS_v2.py:205]: included an if-then-else block

**Review by Gerbenvandervries — APPROVED** [review]

**Pull request merged**